src/package/compressed.rs

```diff
@@ -47,76 +47,112 @@ impl Package {
+            header.set_mode(0o444);
+            header.set_size(contents.len() as u64);
+            archive
+                .append_data(&mut header, name, &contents[..])
+                .into_diagnostic()
+                .wrap_err(miette!("failed to add proto {name:?} to release tar"))?;
+        }
+
+        // Finalize tarball
+        let tar = archive
+            .into_inner()
+            .into_diagnostic()
+            .wrap_err(miette!("failed to assemble tar package"))?;
+
+        // Compress tarball
+        let tgz = Self::compress_tarball(tar)?;
+
+        Ok(Self { manifest, tgz })
+    }
+
+    /// Helper to add a manifest (original or resolved) to the tarball.
+    fn add_manifest_to_archive(
+        archive: &mut tar::Builder<Vec<u8>>,
+        manifest: &Manifest,
+        prefix: Option<&str>,
+        file_name: &str,
+    ) -> miette::Result<()> {
         let manifest_bytes = {
             let as_str: String = manifest
                 .clone()
                 .try_into()
                 .into_diagnostic()
                 .wrap_err(SerializationError(ManagedFile::Manifest))?;
 
-            as_str.into_bytes()
+            // Prepend the prefix if provided
+            let mut result = String::new();
+            if let Some(prefix_text) = prefix {
+                result.push_str(prefix_text);
+            }
+            result.push_str(&as_str);
+            result.into_bytes()
         };
 
         let mut header = tar::Header::new_gnu();
-
         header.set_size(
             manifest_bytes
                 .len()
                 .try_into()
                 .into_diagnostic()
                 .wrap_err(miette!(
-                    "serialized manifest was too large to fit in a tarball"
+                    "serialized manifest `{file_name}` was too large to fit in a tarball"
                 ))?,
         );
-
         header.set_mode(0o444);
 
         archive
-            .append_data(&mut header, MANIFEST_FILE, Cursor::new(manifest_bytes))
-            .into_diagnostic()
-            .wrap_err(miette!("failed to add manifest to release"))?;
-
-        for (name, contents) in &files {
-            let mut header = tar::Header::new_gnu();
-            header.set_mode(0o444);
-            header.set_size(contents.len() as u64);
-            archive
-                .append_data(&mut header, name, &contents[..])
-                .into_diagnostic()
-                .wrap_err(miette!("failed to add proto {name:?} to release tar"))?;
-        }
-
-        let tar = archive
-            .into_inner()
+            .append_data(&mut header, file_name, Cursor::new(manifest_bytes))
             .into_diagnostic()
-            .wrap_err(miette!("failed to assemble tar package"))?;
+            .wrap_err(miette!("failed to add manifest `{file_name}` to release"))?;
+        Ok(())
+    }
 
+    /// Helper to compress the tarball into a `.tgz` file.
+    fn compress_tarball(tar: Vec<u8>) -> miette::Result<Bytes> {
```

Comment:
Can we make a `Tarball` newtype for this?

Reply:
That yields to better separation and easier testability + makes it reusable for other parts of the code

Reply:
Sure: 26395f5ab5ba00862b52918c9d3326ae35fb1a1f. But did you have more in mind that this commit?